servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -32,6 +40,58 @@ data RouteResult a =
   | Route !a
   deriving (Eq, Show, Read, Functor)
 
+instance Applicative RouteResult where
+    pure = return
+    (<*>) = ap
+
+instance Monad RouteResult where
+    return = Route
+    Route a     >>= f = f a
+    Fail e      >>= _ = Fail e
+    FailFatal e >>= _ = FailFatal e
+
+newtype RouteResultT m a = RouteResultT { runRouteResultT :: m (RouteResult a) }
+  deriving (Functor)
+
+-- As we write these instances, we get instances for `DelayedIO` with GND.
```

Comment:
I don't think the comment is necessary. Instances are always good in my opinion.